File: pycodes/hashtable.py
from collections import Counter
from collections import defaultdict
import math

class sloution():
    def _242(self, s, t):
        return Counter(s)==Counter(t)

    def _349(self, nums1, nums2):
        return [x for x in set(nums1) if x in set(nums2)]
    
    def _202(self, n):
    # 快乐数
        num_set=[n]
        while True:
            n = sum([j**2 for j in [n//10**(i)%10 for i in range(int(math.log(n,10))+1)]])
            if n == 1:return True
            elif n in num_set:return False
            num_set.append(n)
    
    # 四数相加
    def _454(self, nums1, nums2, nums3, nums4):
        dic1 = defaultdict(int)
        dic2 = defaultdict(int)
        N = 0
        for (i,j) in zip(nums1,nums2):
            for (m,n) in zip(nums3,nums4):
                dic1[i+m]+=1
                dic2[j+n]+=1
        for key in dic1.keys():
            N+=dic1[key]*dic2[-key]
        return N

    # 赎金信
    def _383(self, ransomNote, magazine):
        a, b = Counter(ransomNote),Counter(magazine)
        for key in a:
            if key not in b:return False
            elif a[key] > b[key]: return False
        return True

    # 三数之和
    """ 
        第一种方法：使用哈希表，注意关键点是去重，以及使用Counter的好处，但是实际上效果很差，其实这道题不适合使用哈希表
    """
    def _15_1(self, nums):
        nums.sort()
        length=len(nums)
        sum_0=[]
        nums_dict=Counter(nums)
        for i in range(length):
            if nums[i]>0:break
            j=i+1
            while j<length:
                if -(nums[i]+nums[j]) in nums_dict:
                    index = nums.index(-(nums[i]+nums[j]))
                    if index+nums_dict[-(nums[i]+nums[j])]-1>j:
                        if [nums[i],nums[j],nums[index]] not in sum_0:
                            sum_0.append([nums[i],nums[j],nums[index]])
                j+=nums_dict[nums[j]]
        return sum_0
    
    """ 第二种思路：使用双指针 """
    def _15_2(self, nums):
        nums.sort()
        print(nums)
        sum_0=[]
        for i in range(len(nums)-2):
            if nums[i]+nums[-1]+nums[-2]<0:continue
            if nums[i]>0:break
            if i>0 and nums[i]==nums[i-1]:continue
            left, right=i+1,len(nums)-1
            while left<right:
                s = nums[i] + nums[left] + nums[right]
                if s>0:
                    right-=1
                    while nums[right]==nums[right+1] and left<right:
                        right-=1
                elif s<0:
                    left+=1
                    while nums[left]==nums[left-1] and left<right:
                        left+=1
                else:
                    sum_0.append([nums[i],nums[left],nums[right]])
                    left+=1
                    right-=1
                    while nums[left]==nums[left-1] and left<right:
                        left+=1
                    while nums[right]==nums[right+1] and left<right:
                        right-=1
        return sum_0
    
    def _19(self, nums, target):
        nums.sort()
        n = len(nums)
        result = []
        for i in range(n):
            if nums[i] > target and nums[i] > 0 and target > 0:# 剪枝（可省）
                break
            if i > 0 and nums[i] == nums[i-1]:# 去重
                continue
            for j in range(i+1, n):
                if nums[i] + nums[j] > target and target > 0: #剪枝（可省）
                    break
                if j > i+1 and nums[j] == nums[j-1]: # 去重
                    continue
                left, right = j+1, n-1
                while left < right:
                    s = nums[i] + nums[j] + nums[left] + nums[right]
                    if s == target:
                        result.append([nums[i], nums[j], nums[left], nums[right]])
                        while left < right and nums[left] == nums[left+1]:
                            left += 1
                        while left < right and nums[right] == nums[right-1]:
                            right -= 1
                        left += 1
                        right -= 1
                    elif s < target:
                        left += 1
                    else:
                        right -= 1
        return result


            


s = sloution()        
print(s._15_2([-1,0,1,2,-1,-4,-2,-3,3,0,4]))

File: cpp/code/hashtable.cpp
#include<iostream>
#include<limits>
#include<math.h>
#include<vector>
#include<unordered_set>
#include <map>
#include <unordered_map>
#include<algorithm>

using namespace std;

class Solution {
public:

    // 242. 有效的字母异位词
    bool isAnagram(string s, string t) {
        int table[26];
        for(char ch:s){
            table[ch - 'a']++;
        }
        for(char ch:s){
            table[ch - 'a']--;
        }
        for(int times: table){
            if(times) return false;
        }
        return true;
    }

    vector<int> intersection(vector<int>& nums1, vector<int>& nums2) {
        unordered_set<int> res;
        unordered_set<int> nums_set(nums1.begin(), nums1.end());
        for(int num: nums2){
            if(nums_set.find(num) != nums_set.end()){
                res.insert(num);
            }
        }
        return vector<int>(res.begin(), res.end());
    }

    // 202 快乐数
    int getsum(int num){
        int sum = 0;
        while(num){
            sum += pow(num % 10, 2);
            num /= 10;
        }
        return sum;
    }
    bool isHappy(int n) {
        unordered_set<int> a;
        while (true){
            if(n == 1){
                return true;
            }
            else{
                if(a.find(n) != a.end()){
                    return false;
                }
                else{
                    a.insert(n);
                }
            }
            n = getsum(n);
        }    
    }


    // 1. 两数之和
    vector<int> twoSum(vector<int>& nums, int target) {
        map<int, int> a;
        int n = nums.size();
        for(int i = 0; i < n; i++){
            auto iter = a.find(target - nums[i]);
            if(iter != a.end()){
                return {iter->second, i};
            }
            a.insert(pair<int, int>(nums[i], i));
        }
        return {};
    }

    // 454 四数相加Ⅱ
    int fourSumCount(vector<int>& A, vector<int>& B, vector<int>& C, vector<int>& D ){
        unordered_map<int, int> umap;
        for(int a:A){
            for(int b: B){
                umap[a+b]++;
            }
        }
        int res = 0;
        for(int c: C){
            for(int d:D){
                if(umap.find(-(c+d)) != umap.end()){
                    res += umap[-(c+d)];
                }
            }
        }
        return res;
    }
    
    // 383 赎金信
    bool canConstruct(string ransomNote, string magazine) {
        unordered_map<char, int> magazine_counter;
        for(char s: magazine){
            magazine_counter[s]++;
        }
        for(char s: ransomNote){
            unordered_map<char, int>::iterator it = magazine_counter.find(s);
            if(it == magazine_counter.end()){
                return false;
            }
            else{
                if(it->second == 0){
                    return false;
                }
                else{
                    it->second--;
                }
            }
        }
        return true;
    }

    // 15 三数之和
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> res;
        sort(nums.begin(), nums.end());
        int n = nums.size();
        for(int i = 0; i < n && nums[i] <= 0; i++){
            if(i > 0 && nums[i] == nums[i-1]){
                continue;
            }
            int left = i + 1, right = n - 1;
            while(left < right){
                if(nums[i]+nums[left]+nums[right] > 0){
                    right--;
                }
                else if(nums[i]+nums[left]+nums[right] < 0){
                    left++;
                }
                else{
                    res.push_back(vector<int>{nums[i], nums[left], nums[right]});
                    while(left < right && nums[right]==nums[right-1]){
                        right--;
                    }
                    while(left < right && nums[left] == nums[left+1]){
                        left++;
                    }
                    left++;
                    right--;
                }
            }
        }
        return res;
    }
};